src/backend/api/meals.js

```diff
@@ -1,48 +1,107 @@
+// @ts-nocheck
 const express = require('express');
 const router = express.Router();
 const knex = require('../database');
 
-router.get('/', async (request, response) => {
+router.get('/', async (req, res) => {
+  let query = knex('meal');
+  if ('maxPrice' in req.query) {
+    query = query.where('price', '<', Number(req.query.maxPrice));
+  }
+
+  // Its does not work
+  if ('availableReservations' in req.query) {
+    query = query.join('reservation', () => {
+      this.on('meal.id', '=', 'reservation.meal_id').orOn(
+        'number_of_guests',
+        '<',
+        'meal.max_reservation'
+      );
+    });
+  }
+
+  if ('title' in req.query && typeof 'title' === 'string') {
+    query = query.whereRaw('title LIKE?', [`%${req.query.title}%`]);
+  }
+
+  if ('dateAfter' in req.query) {
+    query = query.where('created_date', '>', `${req.query.dateAfter}`);
+  }
+
+  if ('dateBefore' in req.query) {
+    query = query.where('created_date', '<', `${req.query.dateBefore}`);
+  }
+  if (req.query.limit) {
+    query = query.where(Number(req.query.limit));
+  }
+
+  if (req.query.sort_key) {
+    if (eq.query.sort_dir) {
+      query = query.orderBy(req.query.sort_key, sort_dir);
+    } else {
+      query = query.orderBy(req.query.sort_key, 'asc');
+    }
+  }
   try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex('meals').select('title');
-    response.json(titles);
+    const mealsRecord = await query;
+    res.json(mealsRecord);
   } catch (error) {
-    throw error;
+    res.status(500).json({ Msg: 'Internal server error' });
   }
 });
-router.get('/future-meals', async (req, res) => {
-  const result = await knex.raw(
-    `SELECT * FROM meal ORDER BY created_date DESC `
-  );
-  return res.json({
-    meals: result[0],
-  });
+
+router.post('/', async (req, res) => {
+  try {
+    const result = await knex.insert(req.body).into('meal');
+    res.json(`ID number ${result} is created`);
+  } catch (error) {
+    res.status(500).json({ Msg: 'Internal server error' });
+  }
 });
-router.get('/past-meals', async (req, res) => {
-  const result = await knex.raw(`SELECT * FROM meal ORDER BY created_date ASC`);
-  return res.json({
-    meals: result[0],
-  });
+
+router.get('/:id', async (req, res) => {
+  try {
+    const paramsValue = req.params.id;
+    const result = await knex('meal').where({ id: paramsValue });
+    res.json(result);
+  } catch (error) {
+    res.status(500).json({ Msg: 'Internal server error' });
+  }
 });
 
-router.get('/all-meals', async (req, res) => {
-  const result = await knex.raw(`SELECT id, title FROM meal`);
-  return res.json({
-    meals: result[0],
-  });
+router.get('/:meal_id', async (req, res) => {
```

Comment:
If the path referring to the title of the meal, it can be something like `meal_title` instead of `meal_id`.